{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "synthetic-context",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import regex\n",
    "import re\n",
    "import re\n",
    "def revers(seq):\n",
    "    codes={\"A\":\"A\", \"C\":\"C\", \"G\":\"G\", \"T\":\"T\", \n",
    "           \"R\":\"[AG]\", \"S\":\"[GC]\", \"B\":\"[CGT]\", \"Y\":\"[CT]\", \n",
    "           \"W\":\"[AT]\", \"D\":\"[AGT]\", \"K\":\"[GT]\", \"N\":\"[ACGT]\", \n",
    "           \"H\":\"[ACT]\", \"M\":\"[AC]\", \"V\":\"[ACG]\", \"X\": \"[ACGT]\"}\n",
    "    \n",
    "    d={'A':'T','G':'C','C':'G','T':'A',\n",
    "      'R':'Y','S':'S','B':'V','Y':'R',\n",
    "       'W':'W','D':'H','K':'M','N':'N',\n",
    "       'H':'D','M':'K','V':'B','X':'X'\n",
    "      }\n",
    "    new=''\n",
    "    for i in seq[::-1]:\n",
    "        new+=d[i]\n",
    "    return new\n",
    "\n",
    "def replace_ambiguity_codes(sequence):\n",
    "    codes={\"A\":\"A\", \"C\":\"C\", \"G\":\"G\", \"T\":\"T\", \n",
    "           \"R\":\"[AG]\", \"S\":\"[GC]\",\"B\":\"[CGT]\", \"Y\":\"[CT]\", \"W\":\"[AT]\",\n",
    "           \"D\":\"[AGT]\", \"K\":\"[GT]\", \"N\":\"[ACGT]\", \"H\":\"[ACT]\", \"M\":\"[AC]\", \"V\":\"[ACG]\", \"X\": \"[ACGT]\"}\n",
    "    regexlist=[]\n",
    "    for base in sequence:\n",
    "        if not base in codes:\n",
    "            print(\"Unidentified nucleotide code in primer:\", base)\n",
    "            sys.exit()\n",
    "        else:\n",
    "            regexlist.append(codes[base])\n",
    "\n",
    "    return ''.join(regexlist)\n",
    "\n",
    "\n",
    "'''\n",
    "In solico PCR\n",
    "def get_target_region(input_fasta,f_primer,r_primer):\n",
    "    f_primer=replace_ambiguity_codes(f_primer)\n",
    "    r_primer=replace_ambiguity_codes(r_primer)\n",
    "    f=open(input_fasta,'r')\n",
    "    fasta_data=f.read().replace('\\r','').split('>')[1:]\n",
    "    f.close()\n",
    "    d1={} # d1: key: ID  value: ATCG\n",
    "    target={}\n",
    "    for each_seq in fasta_data:\n",
    "        d1[each_seq.split(' ')[0]]=each_seq.split('\\n',1)[1].replace('\\n','')\n",
    "    for each_seq in d1:\n",
    "        findall = re.findall(f_primer, d1[each_seq])\n",
    "        if len(findall) == 1:\n",
    "            start=d1[each_seq].split(findall[0])[1]\n",
    "            findall2 = re.findall(r_primer,start)\n",
    "            if len(findall2)==1:\n",
    "                target[each_seq]=start.split(findall2[0])[0]\n",
    "    return len(target),len(d1),len(target)/len(d1),target\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def cov(f,r,d):\n",
    "    result=[]\n",
    "    for i in d:\n",
    "        hit=[]\n",
    "        \n",
    "        for j in d[i]:\n",
    "            forward=re.findall(replace_ambiguity_codes(f),j)\n",
    "            if len(forward) ==1:\n",
    "                reverse=re.findall(replace_ambiguity_codes(revers(r)),j.split(forward[0])[1])\n",
    "                if len(reverse) ==1:\n",
    "                    hit.append(j)\n",
    "        result.append(str((len(hit)/len(d[i]))*100)+'%')\n",
    "        #print(str((len(hit)/len(d[i]))*100)+'%')\n",
    "    return result\n",
    "        #print(len(d[i]))\n",
    "        #print(i)\n",
    "        \n",
    "def cov_e1(f,r,d):\n",
    "    result=[]\n",
    "    for i in d:\n",
    "        hit=[]\n",
    "        \n",
    "        for j in d[i]:\n",
    "            forward=regex.findall('('+replace_ambiguity_codes(f)+'){e<=1}',j)\n",
    "            if len(forward) ==1:\n",
    "                reverse=regex.findall('('+replace_ambiguity_codes(revers(r))+'){e<=1}',j.split(forward[0])[1])\n",
    "                if len(reverse) ==1:\n",
    "                    hit.append(j)\n",
    "        result.append(str((len(hit)/len(d[i]))*100)+'%')\n",
    "        #print(str((len(hit)/len(d[i]))*100)+'%')\n",
    "    return result\n",
    "        #print(len(d[i]))\n",
    "        #print(i)                    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aging-raising",
   "metadata": {},
   "source": [
    "# This script use three database to calculate different primer coverge with zero mismatch and one mismatch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cleared-maria",
   "metadata": {},
   "source": [
    "v1-v3   AGAGTTTGATCCTGGCTCAG    ATTACCGCGGCTGCTGG       27F 534R\n",
    "\n",
    "v4:    GTGYCAGCMGCCGCGGTAA     GGACTACNVGGGTWTCTAAT    515F (Parada)–806R (Apprill)\n",
    "\n",
    "v3-v4   CCTACGGGNGGCWGCAG      GGACTACNVGGGTWTCTAAT    341F 806R\n",
    "\n",
    "v3-v4   CCTACGGGNGGCWGCAG      GACTACHVGGGTATCTAATCC   341F  785R\n",
    "\n",
    "v4-v5   GTGYCAGCMGCCGCGGTAA    CCGYCAATTYMTTTRAGTTT    515F  926R\n",
    "\n",
    "v3-v4  YCTACGGRNGGCHGCAR        RGACTAMNVRGGTMTCTAAT   341F  806R CPR\n",
    "\n",
    "v3-v4  YCTACGGRNGGCHGCAR        RGACTAMNVRGGTHTCTAAT   341F  806R CPR2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "quality-quick",
   "metadata": {},
   "source": [
    "## MiDAS4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "coated-nerve",
   "metadata": {},
   "outputs": [],
   "source": [
    "d={}\n",
    "with open(r'./data/midas4.fasta') as f:\n",
    "    data=f.read().split('>')[1:]\n",
    "    for i in data:\n",
    "        if 'Patesci' in i:\n",
    "            if i.split('\\n')[0].split(',')[2] not in d:\n",
    "                d[i.split('\\n')[0].split(',')[2]] = [i.split('\\n',1)[1].replace('\\n','')]\n",
    "            else:\n",
    "                d[i.split('\\n')[0].split(',')[2]].append(i.split('\\n',1)[1].replace('\\n',''))\n",
    "# merge denovo midas class\n",
    "dd={}\n",
    "dd['denovo_midas_class']=[]\n",
    "for i in d:\n",
    "    if 'midas' in i:\n",
    "        for j in d[i]:\n",
    "            dd['denovo_midas_class'].append(j)\n",
    "    else:\n",
    "        dd[i] = d[i]\n",
    "midas=dd\n",
    "\n",
    "a1=cov('AGAGTTTGATCCTGGCTCAG','ATTACCGCGGCTGCTGG',midas) #27F 534R \n",
    "a2=cov('GTGYCAGCMGCCGCGGTAA','GGACTACNVGGGTWTCTAAT',midas) #515F (Parada)–806R (Apprill)\n",
    "a4=cov('CCTACGGGNGGCWGCAG','GGACTACNVGGGTWTCTAAT',midas) #341F 806R\n",
    "a3=cov('CCTACGGGNGGCWGCAG','GACTACHVGGGTATCTAATCC',midas) #341F  785R\n",
    "a5=cov('GTGYCAGCMGCCGCGGTAA','CCGYCAATTYMTTTRAGTTT',midas) #515F  926R\n",
    "a6=cov('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTMTCTAAT',midas) #CPR v1\n",
    "a7=cov('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTHTCTAAT',midas) #CPR v2\n",
    "\n",
    "b1=cov_e1('AGAGTTTGATCCTGGCTCAG','ATTACCGCGGCTGCTGG',midas) #27F 534R \n",
    "b2=cov_e1('GTGYCAGCMGCCGCGGTAA','GGACTACNVGGGTWTCTAAT',midas) #515F (Prd)–806R (pprill)\n",
    "b4=cov_e1('CCTACGGGNGGCWGCAG','GGACTACNVGGGTWTCTAAT',midas) #341F 806R\n",
    "b3=cov_e1('CCTACGGGNGGCWGCAG','GACTACHVGGGTATCTAATCC',midas) #341F  785R\n",
    "b5=cov_e1('GTGYCAGCMGCCGCGGTAA','CCGYCAATTYMTTTRAGTTT',midas) #515F  926R\n",
    "b6=cov_e1('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTMTCTAAT',midas) #CPR v1\n",
    "b7=cov_e1('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTHTCTAAT',midas) #CPR v2\n",
    "\n",
    "l=[len(midas[i]) for i in midas]\n",
    "m= {'Taxonomy':list(midas.keys()),\n",
    "    'No.Seq':l,\n",
    "    'V1-V3(27F-534R)': a1, \n",
    "    'V4(515F-806R)': a2,\n",
    "    'V3-V4(341F-785R)': a3, \n",
    "    'V3-V4(341F-806R)': a4,\n",
    "    'V4-V5(515F-926R)': a5, \n",
    "    'V3-V4(CPRv1)': a6,\n",
    "       'V3-V4(CPRv2)': a7\n",
    "   }\n",
    "\n",
    "m1= {'Taxonomy':list(midas.keys()),\n",
    "    'No.Seq':l,\n",
    "    'V1-V3(27F-534R)': b1, \n",
    "    'V4(515F-806R)': b2,\n",
    "    'V3-V4(341F-785R)': b3, \n",
    "    'V3-V4(341F-806R)': b4,\n",
    "    'V4-V5(515F-926R)': b5, \n",
    "    'V3-V4(CPRv1)': b6,\n",
    "       'V3-V4(CPRv2)': b7\n",
    "   }\n",
    "\n",
    "import pandas as pd\n",
    "f1=pd.DataFrame(data=m)\n",
    "f1.sort_values(\"No.Seq\",inplace=True,ascending=False)\n",
    "f1.to_csv('./result/midas.csv',index=False)\n",
    "\n",
    "f11=pd.DataFrame(data=m1)\n",
    "f11.sort_values(\"No.Seq\",inplace=True,ascending=False)\n",
    "f11.to_csv('./result/midas_one_mismatch.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "endangered-secretariat",
   "metadata": {},
   "source": [
    "# Silva\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "occasional-cement",
   "metadata": {},
   "outputs": [],
   "source": [
    "silva={}\n",
    "with open(r'./data/silva_CPR.fasta') as f:\n",
    "    data=f.read().split('>')[1:]\n",
    "    for i in data:\n",
    "        if 'Patesci' in i:\n",
    "            if i.split(';')[2] not in silva:\n",
    "                silva[i.split(';')[2]] = [i.split('\\n',1)[1].replace('\\n','').replace('U','T')]\n",
    "            else:\n",
    "                silva[i.split(';')[2]].append(i.split('\\n',1)[1].replace('\\n','').replace('U','T'))\n",
    "\n",
    "                \n",
    "a1=cov('AGAGTTTGATCCTGGCTCAG','ATTACCGCGGCTGCTGG',silva) #27F 534R \n",
    "a2=cov('GTGYCAGCMGCCGCGGTAA','GGACTACNVGGGTWTCTAAT',silva) #515F (Parada)–806R (Apprill)\n",
    "a4=cov('CCTACGGGNGGCWGCAG','GGACTACNVGGGTWTCTAAT',silva) #341F 806R\n",
    "a3=cov('CCTACGGGNGGCWGCAG','GACTACHVGGGTATCTAATCC',silva) #341F  785R\n",
    "a5=cov('GTGYCAGCMGCCGCGGTAA','CCGYCAATTYMTTTRAGTTT',silva) #515F  926R\n",
    "a6=cov('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTMTCTAAT',silva) #CPR v1\n",
    "a7=cov('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTHTCTAAT',silva) #CPR v2\n",
    "\n",
    "b1=cov_e1('AGAGTTTGATCCTGGCTCAG','ATTACCGCGGCTGCTGG',silva) #27F 534R \n",
    "b2=cov_e1('GTGYCAGCMGCCGCGGTAA','GGACTACNVGGGTWTCTAAT',silva) #515F (Prd)–806R (pprill)\n",
    "b4=cov_e1('CCTACGGGNGGCWGCAG','GGACTACNVGGGTWTCTAAT',silva) #341F 806R\n",
    "b3=cov_e1('CCTACGGGNGGCWGCAG','GACTACHVGGGTATCTAATCC',silva) #341F  785R\n",
    "b5=cov_e1('GTGYCAGCMGCCGCGGTAA','CCGYCAATTYMTTTRAGTTT',silva) #515F  926R\n",
    "b6=cov_e1('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTMTCTAAT',silva) #CPR v1\n",
    "b7=cov_e1('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTHTCTAAT',silva) #CPR v2\n",
    "\n",
    "l=[len(silva[i]) for i in silva]\n",
    "m= {'Taxonomy':list(silva.keys()),\n",
    "    'No.Seq':l,\n",
    "    'V1-V3(27F-534R)': a1, \n",
    "    'V4(515F-806R)': a2,\n",
    "    'V3-V4(341F-785R)': a3, \n",
    "    'V3-V4(341F-806R)': a4,\n",
    "    'V4-V5(515F-926R)': a5, \n",
    "    'V3-V4(CPRv1)': a6,\n",
    "       'V3-V4(CPRv2)': a7\n",
    "   }\n",
    "m1= {'Taxonomy':list(silva.keys()),\n",
    "    'No.Seq':l,\n",
    "    'V1-V3(27F-534R)': b1, \n",
    "    'V4(515F-806R)': b2,\n",
    "    'V3-V4(341F-785R)': b3, \n",
    "    'V3-V4(341F-806R)': b4,\n",
    "    'V4-V5(515F-926R)': b5, \n",
    "    'V3-V4(CPRv1)': b6,\n",
    "       'V3-V4(CPRv2)': b7\n",
    "   }\n",
    "import pandas as pd\n",
    "f2=pd.DataFrame(data=m)\n",
    "f2.sort_values(\"No.Seq\",inplace=True,ascending=False)\n",
    "f2.to_csv('./result/silva.csv',index=False)\n",
    "\n",
    "f22=pd.DataFrame(data=m1)\n",
    "f22.sort_values(\"No.Seq\",inplace=True,ascending=False)\n",
    "f22.to_csv('./result/silva_one_mismatch.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "liked-three",
   "metadata": {},
   "source": [
    "# CPR DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bibliographic-cable",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get fasta:\n",
    "\n",
    "d={}\n",
    "with open(r'./data/CPR_DB_16S_withtax.fasta') as f:\n",
    "    data=f.read().split('>')[1:]\n",
    "    for i in data:\n",
    "        if i.split('\\n')[0].split(';')[2] not in d:\n",
    "            d[i.split(';')[2]] = [i.split('\\n',1)[1].replace('\\n','')]\n",
    "        else:\n",
    "            d[i.split(';')[2]].append(i.split('\\n',1)[1].replace('\\n',''))\n",
    "            \n",
    "cpr=d\n",
    "\n",
    "\n",
    "a1=cov('AGAGTTTGATCCTGGCTCAG','ATTACCGCGGCTGCTGG',cpr) #27F 534R \n",
    "a2=cov('GTGYCAGCMGCCGCGGTAA','GGACTACNVGGGTWTCTAAT',cpr) #515F (Parada)–806R (Apprill)\n",
    "a4=cov('CCTACGGGNGGCWGCAG','GGACTACNVGGGTWTCTAAT',cpr) #341F 806R\n",
    "a3=cov('CCTACGGGNGGCWGCAG','GACTACHVGGGTATCTAATCC',cpr) #341F  785R\n",
    "a5=cov('GTGYCAGCMGCCGCGGTAA','CCGYCAATTYMTTTRAGTTT',cpr) #515F  926R\n",
    "a6=cov('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTMTCTAAT',cpr) #CPR v1\n",
    "a7=cov('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTHTCTAAT',cpr) #CPR v2\n",
    "\n",
    "b1=cov_e1('AGAGTTTGATCCTGGCTCAG','ATTACCGCGGCTGCTGG',cpr) #27F 534R \n",
    "b2=cov_e1('GTGYCAGCMGCCGCGGTAA','GGACTACNVGGGTWTCTAAT',cpr) #515F (Prd)–806R (pprill)\n",
    "b4=cov_e1('CCTACGGGNGGCWGCAG','GGACTACNVGGGTWTCTAAT',cpr) #341F 806R\n",
    "b3=cov_e1('CCTACGGGNGGCWGCAG','GACTACHVGGGTATCTAATCC',cpr) #341F  785R\n",
    "b5=cov_e1('GTGYCAGCMGCCGCGGTAA','CCGYCAATTYMTTTRAGTTT',cpr) #515F  926R\n",
    "b6=cov_e1('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTMTCTAAT',cpr) #CPR v1\n",
    "b7=cov_e1('YCTACGGRNGGCHGCAR','RGACTAMNVRGGTHTCTAAT',cpr) #CPR v2\n",
    "\n",
    "\n",
    "l=[len(cpr[i]) for i in cpr]\n",
    "m= {'Taxonomy':list(cpr.keys()),\n",
    "    'No.Seq':l,\n",
    "    'V1-V3(27F-534R)': a1, \n",
    "    'V4(515F-806R)': a2,\n",
    "    'V3-V4(341F-785R)': a3, \n",
    "    'V3-V4(341F-806R)': a4,\n",
    "    'V4-V5(515F-926R)': a5, \n",
    "    'V3-V4(CPRv1)': a6,\n",
    "       'V3-V4(CPRv2)': a7\n",
    "   }\n",
    "m1= {'Taxonomy':list(cpr.keys()),\n",
    "    'No.Seq':l,\n",
    "    'V1-V3(27F-534R)': b1, \n",
    "    'V4(515F-806R)': b2,\n",
    "    'V3-V4(341F-785R)': b3, \n",
    "    'V3-V4(341F-806R)': b4,\n",
    "    'V4-V5(515F-926R)': b5, \n",
    "    'V3-V4(CPRv1)': b6,\n",
    "       'V3-V4(CPRv2)': b7\n",
    "   }\n",
    "import pandas as pd\n",
    "f3=pd.DataFrame(data=m)\n",
    "f3.sort_values(\"No.Seq\",inplace=True,ascending=False)\n",
    "f3.to_csv('./result/cpr.csv',index=False)\n",
    "\n",
    "f33=pd.DataFrame(data=m1)\n",
    "f33.sort_values(\"No.Seq\",inplace=True,ascending=False)\n",
    "f33.to_csv('./result/cpr_one_mismatch.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hollywood-clearance",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
